test(e2e): onboarding tests

QZera opened the pull request:
Prerequisite #169 

Closes #163 

- Tests the welcome screen shown if the rad binary isn't found (e.g. before Radicle is installed)
  - Radicle not being installed is simulated by moving the `.radicle` from it's default location
- Removes `browser.pause` (used on some flaky tests) in favour of using `browser.waitUntil` (asynchronous assertions). This should reduce overall flakiness when adding new tests to the suite.
- Extracts some common helper functions to separate files

QZera committed:
test(e2e): write tests for more onboarding screens

Signed-off-by: Zacharias Fragkiadakis <zacfragkiadakis@gmail.com>

QZera committed:
test(e2e): make some assertions asynchronous

- Refactors some tests in onboarding.spec.ts using `browser.waitUntil` to reduce flakiness
  - These assertions will now poll for the result instead of failing immediately
- Simulates Radicle installation in the onboarding process by renaming the node home directory
- Extracts common queries, actions, and assertions to helper files

Signed-off-by: Zacharias Fragkiadakis <zacfragkiadakis@gmail.com>

QZera commented on test/e2e/helpers/assertions.ts:
```diff
@@ -0,0 +1,24 @@
+ * considered the default state when the workspace is open with a git and rad initialized
+ * repository.
+ */
+export async function expectCliCommandsAndPatchesToBeVisible(workbench: Workbench) {
```
I was thinking of having a more semantic name for this function. Something that expresses that it is the default view of the extension in the sidebar (e.g. `expectDefaultSidebarViewToBeVisible`).
That way, if sections are added, removed in the future, this name remains relevant. Thoughts?

maninak replied:
Good point. While both work, I think the alternative more semantic name would better match how we use the func and scales better over time.

Though it may be more accurate to use a name like `expectStandardSidebarViewsToBeVisible`.

QZera replied:
Renamed as suggested

QZera committed:
refactor(e2e): rename and enhance standard sidebar views assertion

Signed-off-by: Zacharias Fragkiadakis <zacfragkiadakis@gmail.com>

maninak commented on test/e2e/specs/onboarding.spec.ts:
```diff
@@ -1,19 +1,34 @@
   })
 
-  describe('VS Code, *before* the workspace is rad-initialized,', () => {
+  describe('VS Code, *before* Radicle is installed,', () => {
+    const tempPathToNodeHome = `${nodeHomePath}.temp`
+
+    before(async () => {
+      // Simulate radicle "not being installed" by renaming the node home directory
+      await $`mv ${nodeHomePath} ${tempPathToNodeHome}`
+    })
+
+    after(async () => {
+      await $`mv ${tempPathToNodeHome} ${nodeHomePath}`
+      await workbench.executeCommand('Developer: Reload Window')
+    })
```
It seems that (all things considered) we are

1. booting VS Code
2. activating our extension
3. "deleting" the home folder
4. testing
5. "un-deleting" the hove folder

TL;DR: Step 3 must happen before Step 2.

While of course having this is better than not at all, I'm pretty sure that to properly emulate the onboarding flow, the home folder must not exist _before activating the extension_. That is the trickiest part of the detection logic, as we also saw in our code walkthrough, and I'm afraid this approach isn't covering it. And that's arguably the single most critical part of the onboarding flow.

QZera replied:
Great point! You have the order right.

I've moved the "deletion" of the home directory before Step 1 (AFAIK we don't have a good way to do things between Step 1 and Step 2).

There's a small caveat. I had to use wdio's `onWorkerStart` [hook](https://webdriver.io/docs/configuration/#onworkerstart), which is defined in `wdio.conf.ts`. Ideally I would 've liked this logic to be located in the suite itself itself but alas, I don't think there's much we can do about that.

maninak replied:
Good! The hook solution is actually not as bad as I thought it could have been, well done finding it!

I won't be surprised if you come up with an even better approach later too, probably will be something more "out of the box", as things are.

maninak reviewed (changes requested):
Please let me know if the single comment-assumption I've made is inaccurate and I'll insta-approve.

In general, and from a reader's perspective, I like how the testing architecture is shaping up, well done!

Naturally, I'll be able to form a more thorough verdict after I write a few tests of my own.

QZera committed:
test(e2e): move radicle directory **before** launching vscode

- Uses wdio's `onWorkerStart` hook to move the radicle directory before launching vscode
  - This is only done on the onboarding suite worker

Signed-off-by: Zacharias Fragkiadakis <zacfragkiadakis@gmail.com>

maninak reviewed (approved):
Well done, zac, and thank you!
